auditmatica/access_log.py

```diff
@@ -0,0 +1,329 @@
+
+from auditmatica import events
+
+ARCHIVEMATICA_NAME = "Archivematica"
+STORAGE_SERVICE_NAME = "Archivematica Storage Service"
+
+DEFAULT_SS_BASE_URL = "http://127.0.0.1:62081"
+
+UUID_REGEX = "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
+
+ACCESS_LOG_DATE_FORMAT = "%d/%b/%Y:%H:%M:%S %z"
+CEF_DATE_FORMAT = "%b %d %Y %H:%M:%S"
+
+LINE_FORMAT = re.compile(
+    r"""(?P<ip_address>\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}) - (?P<remote_user>.+) \[(?P<datetime>\d{2}\/[a-z]{3}\/\d{4}:\d{2}:\d{2}:\d{2} (\+|\-)\d{4})\] ((\"(?P<method>.+) )(?P<url>.+)(http\/[1-2]\.[0-9]")) (?P<status_code>\d{3}) (?P<bytes_sent>\d+) (["](?P<referrer>(\-)|(.+))["]) (["](?P<user_agent>.+)["]) (["](?P<forwarded_for>.+)["])""",
+    re.IGNORECASE,
+)
+LINE_FORMAT_WITH_USERNAME = re.compile(
+    r"""(?P<ip_address>\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}) - (?P<remote_user>.+) \[(?P<datetime>\d{2}\/[a-z]{3}\/\d{4}:\d{2}:\d{2}:\d{2} (\+|\-)\d{4})\] ((\"(?P<method>.+) )(?P<url>.+)(http\/[1-2]\.[0-9]")) (?P<status_code>\d{3}) (?P<bytes_sent>\d+) (["](?P<referrer>(\-)|(.+))["]) (["](?P<user_agent>.+)["]) (["](?P<forwarded_for>.+)["]) user=(?P<username>.+)""",
+    re.IGNORECASE,
+)
+
+PLACEHOLDER_UUID = "<uuid>"
+PLACEHOLDER_HTTP_PARAMS = "<http_params>"
+PLACEHOLDER_PROCESSING_CONFIG = "<processing_config>"
+PLACEHOLDER_USER_PK = "<user_pk>"
+
+URLS_TO_IGNORE = [
+    # Polling endpoints.
+    "/status/",
+    "/transfer/status/",
+    "/ingest/status/",
+    # DataTable column state management.
+    "/archival-storage/load_state/",
+    "/archival-storage/save_state/",
+    "/backlog/load_state/",
+    "/backlog/save_state/",
+    # Web assets.
+    "/media/",
+    "/static/",
+    "/jsi18n/",
+    "/favicon.ico",
+    # Gearman tasks.
+    "/tasks/",
+    # Calls issued while user browses transfer browser. Since users
+    # cannot access the content from here, we can safely ignore.
+    "/filesystem/children/location",
+    # API calls used internally by Archivematica.
+    "/api/ingest/copy_metadata_files",
+    "/api/administration/dips/atom/levels",
+    "/api/administration/dips/atom/fetch_levels",
+    "/api/v2beta/jobs",
+    "/api/v2beta/task",
+    "/api/filesystem/metadata",
+    "/api/processing-configuration",
+    "/api/v2beta/validate",
+    # The AJAX call issued to populate the Packages tab datatable
+    # can be ignored because the associated user activity (viewing
+    # package information) is already covered.
+    "/packages_ajax",
+]
+
+
+class DateParseError(Exception):
+    pass
+
+
+class LogParseError(Exception):
+    pass
+
+
+def parse_access_log_line(line: str, ss_base_url: str = DEFAULT_SS_BASE_URL):
+    """Parse log line into formatted dictionary.
+
+    If URL is in URLS_TO_IGNORE, return empty dictionary instead.
+
+    :param line: Log line (str)
+    :param ss_base_url: Base URL for Storage Service, as present in
+        nginx access log "referrer" field (str)
+
+    :returns: Parsed and formatted log line (dict)
```

Comment:
I don't know what the 
```suggestion
    :returns: Parsed and formatted log line (dict) or empty dictionary if line cannot be parsed dict).
```
I'm not sure the protocol on the error being raised. Do docstrings sometimes have a `raises` component?

Reply:
It seems them do! I've added them where appropriate, e.g.:
```
:returns: Parsed and formatted log line (dict)
:raises LogParseError: if log line is in unexpected format
```

Reply:
That's great. Thanks Tessa.